mesos-client/src/main/scala/com/mesosphere/mesos/client/CredentialsProvider.scala

```diff
@@ -0,0 +1,97 @@
+package com.mesosphere.mesos.client
+
+import java.net.URL
+import java.time.Instant
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.{BasicHttpCredentials, GenericHttpCredentials, HttpCredentials}
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.ActorMaterializer
+import com.typesafe.scalalogging.StrictLogging
+import de.heikoseeberger.akkahttpplayjson.PlayJsonSupport
+import pdi.jwt.JwtJson
+import pdi.jwt.JwtAlgorithm.RS256
+import play.api.libs.json._
+import play.api.libs.json.Reads._
+
+import scala.async.Async.{async, await}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+
+/**
+  * Base credential provider for [[MesosClient]].
+  */
+trait CredentialsProvider {
+
+  /** @return the next [[HttpCredentials]]. This might be a noop, eg for [[BasicAuthenticationProvider]]. */
+  def nextToken(): Future[HttpCredentials]
+}
+
+/**
+  * Provides DC/OS session tokens as JWT for service accounts.
+  *
+  * @param uid The user id.
+  * @param privateKey The JWT secret.
+  * @param root The DC/OS root URL.
+  * @param system The [[ActorSystem]] used for HTTP requests.
+  * @param materializer The [[ActorMaterializer]] use to unmarshal HTTP responses.
+  * @param context The execution context for transforming the future HTTP response.
+  * @see See [[https://docs.mesosphere.com/1.11/security/ent/iam-api/#obtaining-an-authentication-token]] for details on
+  *      the protocol.
+  */
+case class DcosServiceAccountProvider(uid: String, privateKey: String, root: URL)(
+    implicit system: ActorSystem,
+    materializer: ActorMaterializer,
+    context: ExecutionContext)
+    extends CredentialsProvider
+    with PlayJsonSupport
+    with StrictLogging {
+
+  private def expireIn(duration: Duration): Long = Instant.now.getEpochSecond + duration.toSeconds
+
+  /**
+    * Construct a claim to authenticate and retrieve a session token. The timeout is for this token ''not'' the session
+    * token that we are requesting. That means if ou login request takes more than [[DcosServiceAccountProvider.CLAIM_LIFETIME]]
+    * seconds the authentication will fail.
+    *
+    * @return a claim for the user that expires in twenty seconds.
+    */
+  private def claim: JsObject = Json.obj("uid" -> uid, "exp" -> expireIn(DcosServiceAccountProvider.CLAIM_LIFETIME))
+
+  case class SessionToken(token: String)
+  implicit val sessionRead: Reads[SessionToken] = (JsPath \ "token").read[String].map(SessionToken)
+
+  /** @return a new request for a session token. */
+  def acsTokenRequest: HttpRequest = {
+
+    // Tokenize the claim.
+    val token = JwtJson.encode(claim, privateKey, RS256)
+    val data: String = Json.stringify(Json.obj("uid" -> uid, "token" -> token))
```

Comment:
Hey was this work and code based on https://wiki.mesosphere.com/pages/viewpage.action?pageId=123209735?

I would appreciate the usage of the following terms:
- "DC/OS authentication token"
- "service login token"

for consistency, and for easier code readability. 

What is a session token here?

Reply:
The usage of "authenticate" in the code comments here probably means "log in".

The "login" is the process for exchanging primary credentials (e.g. service account credentials) into secondary credentials (DC/OS authentication token).

Reply:
Thanks for the hint. Please see my follow up https://github.com/mesosphere/usi/pull/99.